Debonding mortar user object

recuero opened the pull request:
Refs. #26283

recuero committed:
Initial CZM code from mortar constraints

recuero committed:
AD-ify cohesive zone modeling tools.

The intent of this update is to use the same tools in the modeling of
cohesive zones when using a "weigthed" or mortar approach to the problem
from the mortar user objects.

recuero committed:
Avoid using a temporary object to initialize TwoMaterialPropertyInterface

recuero committed:
Add displacement jump to mortar CZM object (#26283).

Proper interpolation of rotations and deformation gradients is added.

recuero committed:
Add CZM machinery to user object and apply residual (#26283)

recuero committed:
Add tests for holding current behavior

recuero committed:
Use more libmesh_map_find

recuero committed:
Add more tests

recuero committed:
Refactor CZM mortar user objects

recuero committed:
Add ad-related epsilon in frictional object

recuero committed:
Add some documentation

recuero committed:
Remove trailing whitespace

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1939915/) on 87e6997 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26309/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26309/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fc5457fe08eeae01376e96ef1a3107d27ac539c9`

recuero committed:
Make CIVET compile cleanly

recuero committed:
Add the bilinear mixed mode traction CZM

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1940449/) on b9079aa wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26309/site)

This comment will be updated on new commits.

recuero committed:
Parallel communication of ADReals and ADRankTwoTensors

recuero committed:
Avoid AD singularity

recuero committed:
Add aux kernel to retrieve damage and mode mixity ratio

recuero committed:
Tight tolerances and add output test

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1944513/) on e64ef3c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26309/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26309/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e21a1fbe1cc3120026dc67f4b2e97e90ac732d14`

recuero committed:
Move map query to find the key

- Avoid double calls to CZM computations from base class
- Fix documentation in tests file
- Return on first calls when no map is available
- Do not allow renumbering for parallel tests

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1961689/) on 23f9662 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>0ae736</th>
      <th colspan=4>#26309 23f966</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.07%</td>
      <td>85.07%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>98357</td>
      <td>98360</td>
      <td>+3</td>
      <td>32</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17264</td>
      <td>17263</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26309/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26309/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>0ae736</th>
      <th colspan=4>#26309 23f966</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.15%</td>
      <td>90.05%</td>
      <td>-0.10%</td>
      <td>89.34%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4007</td>
      <td>4507</td>
      <td>+500</td>
      <td>503</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>438</td>
      <td>498</td>
      <td>+60</td>
      <td>60</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26309/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26309/coverage/contact)

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>0ae736</th>
      <th colspan=4>#26309 23f966</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.26%</td>
      <td>88.26%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3932</td>
      <td>3932</td>
      <td>-</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>523</td>
      <td>523</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26309/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26309/coverage/heat_transfer)

### Porous flow

<table>
  <thead>
    <tr>
      <th></th>
      <th>0ae736</th>
      <th colspan=4>#26309 23f966</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.30%</td>
      <td>95.30%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>11132</td>
      <td>11132</td>
      <td>-</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>549</td>
      <td>549</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26309/coverage_diff/porous_flow)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26309/coverage/porous_flow)

### Richards

<table>
  <thead>
    <tr>
      <th></th>
      <th>0ae736</th>
      <th colspan=4>#26309 23f966</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>93.12%</td>
      <td>93.12%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3327</td>
      <td>3327</td>
      <td>-</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>246</td>
      <td>246</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26309/coverage_diff/richards)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26309/coverage/richards)

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>0ae736</th>
      <th colspan=4>#26309 23f966</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.90%</td>
      <td>84.91%</td>
      <td>+0.01%</td>
      <td>95.38%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26939</td>
      <td>26942</td>
      <td>+3</td>
      <td>62</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4792</td>
      <td>4788</td>
      <td>-4</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26309/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26309/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26309/coverage/xfem)

</details>

## Warnings

- `contact` new line coverage rate 89.34% is less than the suggested 90.0%



This comment will be updated on new commits.

recuero committed:
Switch from gap to relative displacement.

For some reason the additional two components of this change, in other words,
the non-zero mixity ratio, seems to cause convergence issues. We must not be getting
the Jacobian right with automatic differentiation since FDP works just fine.

recuero committed:
Various fixes, see below.

- Fix mortar integration of displacement jump vector.
- Fix traction sign (I got it wrong initially)
- Add analysis output for local displacement jump

recuero committed:
Parallel communication of weighted displacement jump

recuero committed:
Delete unused input and regold CZM tests.

recuero committed:
Add heavy tests comparing with interface formulation

recuero committed:
Reduce lines of code for mortar parallel comm.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1989878/) on f31258b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26309/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26309/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ad409367e365d3227cf9c1a0df63b3b9cde9b916`

recuero committed:
Add documentation

dschwen commented on modules/contact/include/utils/MortarContactUtils.h:
```diff
@@ -136,6 +160,202 @@ communicateVelocities(std::unordered_map<const DofObject *, T> & dof_map,
+  TIMPI::push_parallel_vector_data(communicator, push_back_data, sent_back_action_functor);
+}
+
+template <typename T>
```
You have a template parameter here, but the name implies it should always be `ADReal`?

recuero replied:
I have tested `ADReal` and `ADRealVectorValue` for that method. Need to come up with a name that would make sense.

recuero replied:
Maybe `communicateADRealObject`.

dschwen commented on modules/contact/include/utils/MortarContactUtils.h:
```diff
@@ -136,6 +160,202 @@ communicateVelocities(std::unordered_map<const DofObject *, T> & dof_map,
+      if (send_data_back)
+        pid_to_dof_object_for_sending_back[pid].push_back(dof_object);
+
+      for (const auto i : make_range(3))
+        for (const auto j : make_range(3))
+          dof_map_adr2t[dof_object](i, j) += pr.second(i, j);
```
This could be generalized nicely with the new  [`SerialAccess`](https://github.com/idaholab/moose/blob/next/framework/include/utils/SerialAccess.h) stuff ([unit test](https://github.com/idaholab/moose/blob/next/unit/src/SerialAccessTest.C)).

recuero replied:
I'll take a look.

recuero committed:
Change template names.

- Remove Templ and add default template argument for tensorial operations
- Change name of parallel communication templated method

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1992097/) on b614b56 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26309/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26309/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8ac5e2a1afb934d8b46e1d1df9b945d5a0dc26c1`

recuero committed:
Cleanup

dschwen commented on modules/contact/doc/content/source/userobjects/BilinearMixedModeCohesiveZoneModel.md:
```diff
@@ -0,0 +1,153 @@
+### Mode mixity Delamination propagation prediction
+
+The new CZM mortar implementation also accounts for two types of delamination propagation models. (1) A power law-based criterion and (2) The Benzeggagh-Kenane (BK) criterion (see, e.g.,[!cite](benzeggagh1996measurement,bui2011modified)).
+\subsubsection{Power law-based delamination criterion}
```
```suggestion

#### Power law-based delamination criterion

```

dschwen commented on modules/contact/doc/content/source/userobjects/BilinearMixedModeCohesiveZoneModel.md:
```diff
@@ -0,0 +1,153 @@
+    \label{eq:mix_mode_decohesion}
+\end{equation}
+
+### Benzeggagh-Kenane (BK) delamination criterion
```
```suggestion
#### Benzeggagh-Kenane (BK) delamination criterion
```

dschwen commented on modules/contact/doc/content/source/userobjects/BilinearMixedModeCohesiveZoneModel.md:
```diff
@@ -0,0 +1,153 @@
+
+### Constitutive model
+
+The CZM constitutive models relate the ``weighted'' interface jump components to the interface traction in the local frame of reference. The traction, in the local frame of reference\footnote{The local frame of reference is given by the secondary surface employing an interpolated normal vector.}, can be expressed as:
```
```suggestion
The CZM constitutive models relate the weighted interface jump components to the interface traction in the local frame of reference. The traction, in the local frame of reference\footnote{The local frame of reference is given by the secondary surface employing an interpolated normal vector.}, can be expressed as:
```

dschwen replied:
footnote

dschwen commented on modules/contact/doc/content/source/userobjects/BilinearMixedModeCohesiveZoneModel.md:
```diff
@@ -0,0 +1,153 @@
+\label{eq:constitutive}
+\end{equation}
+where $\tilde{\delta}_{\text{active}}$ represents the CZM weighted interface jump with shear components, i.e. it applies shear stiffness, and normal stiffness only if a gap exists. $\tilde{\delta}_{\text{inactive}}$ applies a normal traction that prevents interpenetration. A few notes about Equation~(\ref{eq:constitutive}):
+\begin{itemize}
+    \item The action of $\delta_{\text{inactive}}$ can potentially be replaced with (mortar) mechanical contact enforcement.
+    \item The local traction $\tau_{l}$ is computed nodally with ``weighted'' quantities, but is interpolated and applied on the primary and secondary surfaces using Lagrange interpolation functions.
+    \item The constitutive model can be expanded as needed in the user object.
+\end{itemize}
```
markdown

hugary1995 commented:
Very exciting capability! How are we handling stateful data on CZM interface?

dschwen commented on modules/contact/include/auxkernels/CohesiveZoneMortarUserObjectAux.h:
```diff
@@ -0,0 +1,68 @@
+    MODE_MIXITY_RATIO,
+    COHESIVE_DAMAGE,
+    LOCAL_NORMAL_JUMP,
+    LOCAL_TANGENTIAL_JUMP
```
Note: first direction only

dschwen commented on modules/contact/include/userobjects/BilinearMixedModeCohesiveZoneModel.h:
```diff
@@ -0,0 +1,191 @@
+  /// Compute global traction for mortar application
+  virtual void computeGlobalTraction(const Node * const node) override;
+
+  /// Normalize rank two tensor mortar quantities (remove mortar integral scaling)
+  virtual ADRankTwoTensor
+  normalizeRankTwoTensorQuantity(const std::unordered_map<const DofObject *, ADRankTwoTensor> & map,
+                                 const Node * const node);
+
+  /// Normalize real mortar quantities (remove mortar integral scaling)
+  virtual ADReal normalizeRealQuantity(const std::unordered_map<const DofObject *, ADReal> & map,
+                                       const Node * const node);
+
+  /// Normalize real vector mortar quantities (remove mortar integral scaling)
+  virtual ADRealVectorValue
+  normalizeVectorQuantity(const std::unordered_map<const DofObject *, ADRealVectorValue> & map,
+                          const Node * const node);
```
not: we could prbly template this

dschwen commented on modules/contact/include/userobjects/BilinearMixedModeCohesiveZoneModel.h:
```diff
@@ -0,0 +1,191 @@
+  normalizeVectorQuantity(const std::unordered_map<const DofObject *, ADRealVectorValue> & map,
+                          const Node * const node);
+
+  /// The stress tensor on the interface
+  // const ADMaterialProperty<RankTwoTensor> & _stress;
+
```
```suggestion
```

dschwen commented on modules/contact/include/userobjects/BilinearMixedModeCohesiveZoneModel.h:
```diff
@@ -0,0 +1,191 @@
+  const MaterialProperty<Real> & _normal_strength;
+
+  /// The shear strength material property
+  const MaterialProperty<Real> & _shear_strength;
```
note: `_shear_strength` = T = T<sub>1</sub> = T<sub>2</sub>

dschwen commented on modules/contact/src/userobjects/BilinearMixedModeCohesiveZoneModel.C:
```diff
@@ -0,0 +1,562 @@
+const ADVariableValue &
+BilinearMixedModeCohesiveZoneModel::czmGlobalTraction(const unsigned int i) const
+{
+  if (i >= 3)
+    mooseError("Internal error in czmGlobalTraction.");
+
+  if (i == 0)
+    return _czm_interpolated_traction_x;
+  else if (i == 1)
+    return _czm_interpolated_traction_y;
+  else
+    return _czm_interpolated_traction_z;
```
note: use vector of components

dschwen commented on modules/contact/include/userobjects/PenaltySimpleCohesiveZoneModel.h:
```diff
@@ -0,0 +1,86 @@
+  /// Map from degree of freedom to czm normal traction
+  std::unordered_map<const DofObject *, ADReal> _dof_to_czm_normal_traction;
+
+  /// The first frictional contact pressure on the mortar segment quadrature points
+  ADVariableValue _frictional_contact_traction_one;
+
+  /// The second frictional contact pressure on the mortar segment quadrature points
+  ADVariableValue _frictional_contact_traction_two;
```
note: not tested. Maybe we want to inherit from the proper base class instead

dschwen replied:
below as well

dschwen commented on modules/contact/src/userobjects/WeightedGapUserObject.C:
```diff
@@ -107,6 +115,9 @@ WeightedGapUserObject::computeQpProperties()
   _qp_factor = _JxW_msm[_qp] * _coord[_qp];
+
+  // Call child objects  QpProperties
+  computeQpPropertiesLocal();
```
Use different pattern, where the child calls the parent class(es)'(s) compute method first

dschwen commented on modules/contact/src/userobjects/PenaltyFrictionUserObject.C:
```diff
@@ -264,8 +266,13 @@ PenaltyFrictionUserObject::reinit()
-      if (penalty_friction * slip_distance.norm() >
-          0.4 * _friction_coefficient * std::abs(normal_pressure))
+
+      const auto slip_metric = std::abs(MetaPhysicL::raw_value(slip_distance).cwiseAbs()(0)) +
+                               std::abs(MetaPhysicL::raw_value(slip_distance).cwiseAbs()(1));
```
Double abs

dschwen commented on modules/contact/src/userobjects/PenaltySimpleCohesiveZoneModel.C:
```diff
@@ -0,0 +1,263 @@
+    _penalty_friction(isParamValid("penalty_friction") ? getParam<Real>("penalty_friction")
+                                                       : getParam<Real>("penalty")),
+    _friction_coefficient(getParam<Real>("friction_coefficient")),
+    _epsilon_tolerance(1.0e-40)
```
note: think about global constant for this

recuero committed:
Address Daniel's comments and cleanup

recuero commented:
@hugary1995,

> How are we handling stateful data on CZM interface?
 
 The states are associated with secondary surface nodes. I.e., mortar quantities are computed and assigned to nodes, which may have old states if required (e.g. damage). This approach is akin to the mortar mechanical contact that we have. I thought proceeding this way would be an easy extension of mechanical contact to a mortar CZM interface. If properties at quadrature points needed to be evaluated (which is not required here as this PR stands), then the work that @dschwen did here https://github.com/idaholab/moose/pull/26054 would be necessary. Those stateful material properties are also necessary for mortar thermal contact if old states are to be used.

hugary1995 commented:
It would be nice if we permit material storage on _stationary_ mortar interfaces. @lindsayad @recuero What's your opinion on this?

recuero commented:
A summary of items that can be revamped (from Daniel's comments/notes):

 - The mortar auxiliary kernel created here should also include the '3D' tangential direction.
 - A constant for eliminating AD singularities can be created (libMesh `TOLERANCE` is too large for that application)
 - `PenaltySimpleCohesiveModel` could inherit directly from the frictional object. AL should be tested in conjunction with CZM capabilities.
 - Shear strengths values must be isotropic for now. That is, the two tangential directions in 3D behave similarly. Considering anisotropy requires changing the code.

hugary1995 commented:
At lease for the traditional CZM, the interface doesn't need to move/change. If we can enable material storage for mortar qps, we don't need to rely on NPR to project and interpolate properties back and forth. We've also discussed earlier that some quantities, e.g. orientation, rotation, will require additional effort for projection to work properly.

recuero commented:
Alex will know better, but that may require significant framework development. With the approach in this PR, you may achieve satisfactory results with the changing mortar interface at the expense of having a biased formulation.

hugary1995 commented:
> With the approach in this PR, you may achieve satisfactory results with the changing mortar interface at the expense of having a biased formulation.

I have no doubt on that lol.

lindsayad commented:
With @dschwen's recent #26054 I would be reticent to add an additional stateful mechanism until there is some proof that what we have already is unsatisfactory. Note that #26054 doesn't require projection into a Lagrange basis. You can choose a discontinuous basis as well; perhaps that could better minimize projection error. If you did bring me proof that this is unsatisfactory, then I would definitely want to help you with a qp-based storage mechanism on undisplaced meshes

hugary1995 commented:
It can be proved by pure math that the linear interpolation of orientation is wrong (not just off). It's a matter of "when we need material models that require orientation as a stateful property" rather than "when we see unsatisfactory results".

hugary1995 commented:
Bottom line is I'm fine with the current approach. It's just not future-proof.

lindsayad commented:
what is the orientation a function of? Does it relate to the mapping of the element?

recuero commented:
> It can be proved by pure math that the linear interpolation of orientation is wrong (not just off).

True. But couldn't other interpolation strategies with this approach? We do need specific interpolation for applying tractions, but other quantities could be interpolated differently as far as I can tell.

hugary1995 commented:
There are various of interpolating orientation. Arguably the most straightforward method is to solve a singular value problem (N by 4 weighted quaternions). The singular vector corresponding to the largest singular value is the interpolated quaternion. I don't know how difficult that is to implement in the framework of #26054

hugary1995 commented:
> what is the orientation a function of? Does it relate to the mapping of the element?

My bad for not including context. I was referring to orientation as a general material property or model parameter, not related to finite element stuff.

recuero commented:
That's an option. For the nodal approach used here you can build rotation matrices with interpolated (mortar) normal vectors, see https://github.com/recuero/moose/blob/7a393edb9fdd5a2162b785fb92ff1fe203781e30/modules/contact/src/userobjects/BilinearMixedModeCohesiveZoneModel.C#L221.

hugary1995 commented:
Yes, that will definitely work. But it's different from the scenario I was trying to describe.

Maybe think about it this way: Some users (e.g. me) want to use debonding mortar userobject between grain boundaries. The debonding traction separation law depends on the grain orientation of the neighboring grains. That grain orientation is truly a material property that can vary spatially, independent of any mortar geometrical quantity. When we integrate the traction separation law incrementally, we may need the grain orientation to be stateful, and so the current NPR approach (i.e. projecting back and forth onto FE basis) probably doesn't work.

lindsayad commented:
and the spatial variation is some arbitrary function that cannot be represented by any FE basis that we have (?)

hugary1995 commented:
Quaternion is just a parameterization of SO(3), there are other parameterizations. Any polynomial interpolation of 3- or 4-parameter parameterization of SO(3) will only work for a limited space. So to confirm your statement:

"the spatial variation _can be_ some arbitrary function that cannot be (exactly) represented by any FE basis that we have"
